presto-main/src/main/java/com/facebook/presto/operator/aggregation/AreaUnderRocCurveAggregation.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.operator.aggregation.state.AreaUnderRocCurveState;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.AggregationState;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import it.unimi.dsi.fastutil.ints.AbstractIntComparator;
+import it.unimi.dsi.fastutil.ints.IntArrays;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+
+@AggregationFunction("area_under_roc_curve")
+public final class AreaUnderRocCurveAggregation
+{
+    private AreaUnderRocCurveAggregation() {}
+
+    @InputFunction
+    public static void input(@AggregationState AreaUnderRocCurveState state, @SqlType(StandardTypes.BOOLEAN) boolean label, @SqlType(StandardTypes.DOUBLE) double score)
+    {
+        MultiKeyValuePairs pairs = state.get();
```

Comment:
Using MultiKeyValuePairs is overkill and likely to have sub-optimal performance. Moreover, having a MultiKeyValuePairs instance per group will cause significant overhead.

For the grouped state, the data structure should be flattened so that all groups and values are stored in a single set of parallel arrays (similar to how https://github.com/prestodb/presto/pull/8918 does it)

Reply:
To elaborate, something like this (effectively, a set of interleaved linked lists)

```
// one entry per group. Each entry is the index of the first element of the group 
// in the labels/scores/nextLinks arrays
int[] heads; 

// one entry per score/label entry. 
// The value in nextLinks contains the index of the next value in the chain. A value of -1 
// indicates it's the last value for the group
boolean[] labels;
double[] scores;
int[] nextLinks;

// the index of the next free element in the labels/scores/nextLinks arrays. This is needed to be able know 
// where to continue adding elements when after the arrays are resized.
int nextFree;
```

Adding an element to a group would be something like:

```
labels[nextFree] = label;
scores[nextFree] = score;
nextLinks[nextFree] = heads[group];

heads[group] = nextFree;
nextFree++;
```

Reply:
Thank you for providing the suggestion. I would look into how #8918 handles grouped state and get back later.